src/model/Model_CurrencyHistory.cpp

```diff
@@ -85,41 +85,68 @@ int Model_CurrencyHistory::addUpdate(const int& currencyID, const wxDate& date,
 /** Return the rate for a specific currency in a specific day*/
 double Model_CurrencyHistory::getDayRate(const int& currencyID, const wxString& DateISO)
 {
-    if (currencyID == Model_Currency::GetBaseCurrency()->CURRENCYID)
+    if (currencyID == Model_Currency::GetBaseCurrency()->CURRENCYID || currencyID == -1)
         return 1;
     
     wxDateTime Date;
     Date.ParseDate(DateISO);
     Model_CurrencyHistory::Data_Set Data = Model_CurrencyHistory::instance().find(Model_CurrencyHistory::CURRENCYID(currencyID),Model_CurrencyHistory::CURRDATE(Date));
 
     if (!Data.empty())
+    {
+        //Rate found for specified day
         return Data.back().CURRVALUE;
+    }
+    else if (Model_CurrencyHistory::instance().find(Model_CurrencyHistory::CURRENCYID(currencyID)).size() > 0)
+    {
+        //Rate not found for specified day
+        //Custom query requested to speed-up performances, no way to obtain it with our ORM
+        wxDateTime dFuture, dPast, dNearest;
+
+        const wxString sqlPast = wxString::Format("SELECT MAX(currdate) FROM CURRENCYHISTORY WHERE currencyid = '%i' AND currdate <= '%s';", currencyID, DateISO);
+        wxSQLite3ResultSet rsPast = Model_CurrencyHistory::instance().db_->ExecuteQuery(sqlPast);
+        while (rsPast.NextRow())
+        {
+            dPast.ParseDate(rsPast.GetAsString(0));
+        }
+
+        const wxString sqlFuture = wxString::Format("SELECT MIN(currdate) FROM CURRENCYHISTORY WHERE currencyid = '%i' AND currdate >= '%s';", currencyID, DateISO);
+        wxSQLite3ResultSet rsFuture = Model_CurrencyHistory::instance().db_->ExecuteQuery(sqlFuture);
+        while (rsFuture.NextRow())
+        {
+            dFuture.ParseDate(rsFuture.GetAsString(0));
+        }
+
+        if (dPast.IsValid() && dFuture.IsValid())
+        {
+            const wxTimeSpan spanPast = Date.Subtract(dPast);
+            const wxTimeSpan spanFuture = dFuture.Subtract(Date);
+
+            dNearest = spanPast <= spanFuture ? dPast : dFuture;
```

Comment:
This should be ok for currencies - they should be continuous and we can do linear interpolation here.
But for shares (not updated every day) it should be better to use dPast always.

Reply:
What do you mean for shares? If you refer to shares value, than there is StockHistory, if you refer to conversion of shares value, then should be fine because it will try to find nearest conversion for the date of share value, isn't it?

Reply:
Yes, I'm writing about StockHistory here, for the future...